### PR TITLE
Improve reset code and make it so VRTrackers can reset

### PR DESCRIPTION
**Pull request by Louka3000**
- Improves some reset code and make sure none of the fix___ methods are modifying the tracker's rotation (which they were, and it caused trackers to jump for a frame sometimes when resetting)
- Make it so VRTrackers (Vive trackers) can align to HMD when resetting and quick resetting. First step in supporting Vive trackers... this will already make them more usable!

### CODE REVIEW

File: server/src/main/java/dev/slimevr/vr/trackers/VRTracker.java [hunk 1]
```diff
@@ -23,6 +27,54 @@ public VRTracker(int id, String name, boolean hasRotation, boolean hasPosition)
 		super(id, name, name, hasRotation, hasPosition, null);
 	}
 
+	@Override
```
Comment (ButterscotchV) [inline server/src/main/java/dev/slimevr/vr/trackers/VRTracker.java:30]: I'd love to at least have some kinda note about this code duplication for future reference
```suggestion
	// TODO: Reduce code duplication from IMUTracker.java
	@Override
```

File: server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java [hunk 2]
```diff
@@ -406,8 +406,7 @@ public void resetFull(Quaternion reference) {
 		fixGyroscope(getMountedAdjustedRotation());
 		fixAttachment(getMountedAdjustedRotation());
 		makeIdentityAdjustmentQuatsFull();
-		fixYaw(reference);
-		calibrateMag();
```
Comment (ButterscotchV) [inline server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java:410]: Why is `calibrateMag();` removed here?

**Review by ButterscotchV — CHANGES_REQUESTED** [review]

File: server/src/main/java/dev/slimevr/vr/trackers/VRTracker.java (hunk 1 above)
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/vr/trackers/VRTracker.java:30]: Well, I can add it, but we plan to refactor all of the classes into a single one, like discussed with Futura

File: server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java (hunk 2 above)
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java:410]: git conflicts ;-;

File: server/src/main/java/dev/slimevr/vr/trackers/VRTracker.java (hunk 1 above)
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/vr/trackers/VRTracker.java:30]: Added both notes in the end

**Review by ButterscotchV — APPROVED** [review]

File: server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java [hunk 3]
```diff
@@ -490,40 +492,40 @@ public void resetMounting(boolean reverseYaw) {
 		mountRotFix.set(buffer);
 
 		// Get the difference from the last adjustment
-		buffer.multLocal(lastRotAdjust.inverseLocal());
+		buffer.multLocal(lastRotAdjust.inverse());
 		// Apply the yaw rotation difference to the yaw fix quaternion
-		yawFix.multLocal(buffer.inverseLocal());
+		yawFix.multLocal(buffer.inverse());
```
Comment (Eirenliel) [inline server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java:497]: It's buffer variable, do not make unnecessary copies, it should be inverseLocal()

File: server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java [hunk 4]
```diff
@@ -490,40 +492,40 @@ public void resetMounting(boolean reverseYaw) {
...
-		Quaternion sensorRotation = getMountedAdjustedRotation();
+		sensorRotation = sensorRotation.clone();
 		gyroFix.mult(sensorRotation, sensorRotation);
 		sensorRotation.multLocal(attachmentFix);
 		sensorRotation.multLocal(mountRotFix);
 
 		sensorRotation.fromAngles(0, sensorRotation.getYaw(), 0);
 
-		yawFix.set(sensorRotation).inverseLocal().multLocal(targetRotation);
+		yawFix.set(sensorRotation.inverse().mult(reference));
```
Comment (Eirenliel) [inline server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java:512]: Same here, this cloning is not needed.

File: server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java [hunk 5]
```diff
@@ -490,40 +492,40 @@ public void resetMounting(boolean reverseYaw) {
...
 	}
 
 	private void calibrateMag() {
 		if (magCalibrationStatus >= CalibrationAccuracy.HIGH.status) {
 			magnetometerCalibrated = true;
 			// During calibration set correction to match magnetometer readings
 			// TODO : Correct only yaw
-			correction.set(rotQuaternion).inverseLocal().multLocal(rotMagQuaternion);
+			correction.set(rotQuaternion.inverse().mult(rotMagQuaternion));
```
Comment (Eirenliel) [inline server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java:520]: It's specifically ordered to first set, then inverseLocal and multLocal so nothing outside is modified and clonning is not needed.

File: server/src/main/java/dev/slimevr/vr/trackers/VRTracker.java [hunk 6]
```diff
@@ -23,6 +27,56 @@ public VRTracker(int id, String name, boolean hasRotation, boolean hasPosition)
...
+	public void resetYaw(Quaternion reference) {
+		fixYaw(rotation, reference);
+	}
+
+	private void fixAttachment(Quaternion sensorRotation, Quaternion reference) {
+		reference = reference.clone();
+		mountFix.fromAngles(0, reference.getYaw(), 0);
+
+		sensorRotation = sensorRotation.clone();
+		attachmentFix.set(sensorRotation.inverse());
```
Comment (Eirenliel) [inline server/src/main/java/dev/slimevr/vr/trackers/VRTracker.java:63]: Do not cloe and then inverse. Either inverse() or clone and inverseLocal(). Also modifying argument variable is bad tone.

File: server/src/main/java/dev/slimevr/vr/trackers/VRTracker.java [hunk 7]
```diff
@@ -23,6 +27,56 @@ public VRTracker(int id, String name, boolean hasRotation, boolean hasPosition)
...
+	}
+
+	@Override
+	public void resetYaw(Quaternion reference) {
+		fixYaw(rotation, reference);
+	}
+
+	private void fixAttachment(Quaternion sensorRotation, Quaternion reference) {
+		reference = reference.clone();
+		mountFix.fromAngles(0, reference.getYaw(), 0);
```
Comment (Eirenliel) [inline server/src/main/java/dev/slimevr/vr/trackers/VRTracker.java:60]: why would you just clone it here, getYaw() doesn't modify the object!

File: server/src/main/java/dev/slimevr/vr/trackers/VRTracker.java [hunk 8]
```diff
@@ -23,6 +27,56 @@ public VRTracker(int id, String name, boolean hasRotation, boolean hasPosition)
...
+
+		sensorRotation = sensorRotation.clone();
+		attachmentFix.set(sensorRotation.inverse());
+	}
+
+	private void fixYaw(Quaternion sensorRotation, Quaternion reference) {
+		// Use only yaw HMD rotation
+		reference = reference.clone();
+		reference.fromAngles(0, reference.getYaw(), 0);
```
Comment (Eirenliel) [inline server/src/main/java/dev/slimevr/vr/trackers/VRTracker.java:69]: same here and below...

**Review by Eirenliel — CHANGES_REQUESTED** [review]
This is not improved reset code, you added unnecessary cloning everywhere...

**Comment by Louka3000** [comment]
Somehow thought inverseLocal created a new Quaternion and inverse didn’t, but it’s the opposite ;-;

File: server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java [hunk 9]
```diff
@@ -456,13 +456,15 @@ protected void adjustToIdentity(Quaternion store) {
 	}
 
 	private void fixGyroscope(Quaternion sensorRotation) {
+		sensorRotation = sensorRotation.clone();
```
Comment (Eirenliel) [inline server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java:459]:
```suggestion
```

**Review by Eirenliel — CHANGES_REQUESTED** [review]

File: server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java (hunk 9 above)
Comment (Louka3000) [reply server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java:459]: The line right below that modifies it, so it needs the clone()
		`sensorRotation.fromAngles(0, sensorRotation.getYaw(), 0);`

Comment (Louka3000) [reply server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java:459]: Could do `sensorRotation = new Quaternion().fromAngles(0, sensorRotation.getYaw(), 0);` if you'd prefer a single line ig

Comment (Eirenliel) [reply server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java:459]: Then replace inverse with inverseLocal below

Comment (Louka3000) [reply server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java:459]: :+1:

File: server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java [hunk 10]
```diff
@@ -456,13 +456,15 @@ protected void adjustToIdentity(Quaternion store) {
...
 		sensorRotation.fromAngles(0, sensorRotation.getYaw(), 0);
 		gyroFix.set(sensorRotation).inverseLocal();
```
Comment (Eirenliel) [inline server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java:461]:
```suggestion
		gyroFix.set(sensorRotation.inverseLocal());
```
If you want to be consistent with other changes.

File: server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java [hunk 11]
```diff
@@ -495,19 +497,19 @@ public void resetMounting(boolean reverseYaw) {
...
-		Quaternion sensorRotation = getMountedAdjustedRotation();
+		sensorRotation = sensorRotation.clone();
 		gyroFix.mult(sensorRotation, sensorRotation);
 		sensorRotation.multLocal(attachmentFix);
 		sensorRotation.multLocal(mountRotFix);
 
 		sensorRotation.fromAngles(0, sensorRotation.getYaw(), 0);
 
-		yawFix.set(sensorRotation).inverseLocal().multLocal(targetRotation);
+		yawFix.set(sensorRotation).inverseLocal().multLocal(reference);
```
Comment (Eirenliel) [inline server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java:512]:
```suggestion
		yawFix.set(sensorRotation.inverseLocal().multLocal(reference));
```

**Review by Eirenliel — CHANGES_REQUESTED** [review]
Some suggestions for consistency.

**Review by Eirenliel — APPROVED** [review]

**Pull request merged**